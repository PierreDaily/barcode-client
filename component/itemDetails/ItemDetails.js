import React, { useState, useEffect } from "react";
import { Formik } from "formik"
import * as yup from 'yup';
import {
  Text,
  Button,
  View,
  TextInput,
  StyleSheet,
  Picker
} from "react-native";
import API from "../api/api";

const getBrandList = () => API.get(`/brand`);

const yupSchema = yup.object().shape({
  brand: yup.string().required(),
  name: yup
    .string()
    .required()
});

export default ItemDetails = ({ route, navigation }) => {
  const [brand, setBrand] = useState(null);
  const [brandList, setBrandList] = useState([]);

  useEffect(() => {
    getBrandList()
      .then(({ data }) => {
        setBrandList(data);
        setBrand(data[0].id);
      })
      .catch(err => alert(err));
  }, []);

  return (
    <View>
      <Text>Barcode: {route.params.barcode}</Text>
      {brand && (<Formik
        initialValues={{ name: '' }}
        validateOnChange={false}
        validationSchema={yupSchema}
        onSubmit={(values) => navigation.navigate("Item-photo", { ...route.params, brand: values.brand, name: values.name })}
      >
        {({ errors, handleSubmit, setFieldValue, values }) => <View>
          <Text>Name</Text>
          <TextInput
            style={[{ height: 40, borderColor: "gray", borderWidth: 1 }, errors.name && styles.error]}
            onChangeText={text => setFieldValue('name', text)}
            value={values.name}
          />

          <Text>Brand</Text>
          <Picker
            selectedValue={values.brand}
            style={[{ height: 50, width: 200 }, errors.brand && styles.error]}
            onValueChange={(itemValue, itemIndex) => setFieldValue('brand', itemValue)}
          >
            {brandList.length > 0 &&
              brandList.map(brand => (
                <Picker.Item label={brand.name} key={brand.id} value={brand.id} />
              ))}
          </Picker>
          <Button
            title="Submit"
            onPress={handleSubmit}
          />
        </View>}
      </Formik>)}
    </View >
  );
};

const styles = StyleSheet.create({
  error: {
    backgroundColor: "red"
  },
});
