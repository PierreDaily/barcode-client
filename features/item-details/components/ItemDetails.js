import React from "react";
import PropTypes from "prop-types";
import { Formik } from "formik";
import * as yup from "yup";
import { View } from "react-native";
import { TextInput } from "react-native-paper";
import { Button } from "../../ui/components";
import styles from "../styles";
import Logo from "../../../assets/img/green-barcode.svg";

const yupSchema = yup.object().shape({
  brandId: yup.number().required(),
  brandName: yup.string().required(),
  name: yup.string().required()
});

const ItemDetails = ({ route, navigation }) => {
  const {
    params: { brandId, brandName, barcode, type }
  } = route;

  return (
    <Formik
      initialValues={{
        name: ""
      }}
      validateOnChange={false}
      validate={async values => {
        const errors = {};
        try {
          await yupSchema.validate({
            brandId,
            brandName,
            name: values.name
          });
        } catch (err) {
          errors[err.path] = err.message;
        }

        return errors;
      }}
      onSubmit={values =>
        navigation.navigate("Item-photo", {
          barcode,
          brand: brandId,
          name: values.name,
          type
        })
      }
      testID="form"
    >
      {({ errors, handleSubmit, setFieldValue, values }) => (
        <View style={styles.container}>
          <Logo style={styles.logo} />
          <View>
            <TextInput
              error={errors.name}
              label="Product name"
              onChangeText={text => setFieldValue("name", text)}
              style={styles.nameInput}
              testID="textInput"
              value={values.name}
            />

            <TextInput
              error={errors.brandName}
              label="Brand Name"
              style={styles.brandInput}
              value={brandName || ""}
              onFocus={() =>
                navigation.navigate("Search", {
                  barcode,
                  type
                })
              }
              testID="search"
            />
          </View>
          <Button
            mode="contained"
            onPress={handleSubmit}
            style={styles.btn}
            testID="submit"
          >
            Next
          </Button>
        </View>
      )}
    </Formik>
  );
};

ItemDetails.propTypes = {
  navigation: PropTypes.shape({
    navigate: PropTypes.func.isRequired
  }),
  route: PropTypes.shape({ params: PropTypes.object })
};

export { ItemDetails };
