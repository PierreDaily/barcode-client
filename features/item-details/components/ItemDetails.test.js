import React from "react";
import {
  act,
  cleanup,
  render,
  fireEvent,
  flushMicrotasksQueue
} from "react-native-testing-library";
import { ItemDetails } from "./ItemDetails";
import renderer from "react-test-renderer";

jest.mock("../../../assets/img/green-barcode.svg", () => ({
  __esModule: true,
  default: "Logo"
}));

let minProps;

beforeEach(() => {
  minProps = {
    navigation: { navigate: jest.fn() },
    route: { params: { barcode: "123456789", type: "EAN" } }
  };
});

afterEach(() => {
  jest.clearAllMocks();
  cleanup();
});

test("renders correctly", async () => {
  const tree = renderer.create(<ItemDetails {...minProps} />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("should submit the form with the correct informations", async () => {
  const itemName = "random item";
  const props = {
    ...minProps,
    route: {
      params: { ...minProps.route.params, brandName: "Brand Name", brandId: 4 }
    }
  };
  const { getByTestId } = render(<ItemDetails {...props} />);

  const {
    route: {
      params: { barcode, brandId, type }
    }
  } = props;
  const nextScreen = "Item-photo";

  expect(getByTestId("form")).toBeTruthy();

  fireEvent.changeText(getByTestId("textInput"), itemName);
  await act(async () => fireEvent(getByTestId("submit"), "onPress"));

  expect(props.navigation.navigate).toHaveBeenCalledWith(nextScreen, {
    barcode,
    brand: brandId,
    name: itemName,
    type
  });
});

test("shouldn't submit the form because informations are missing", async () => {
  const itemName = "random item";
  const { getByTestId } = render(<ItemDetails {...minProps} />);

  expect(getByTestId("form")).toBeTruthy();

  fireEvent.changeText(getByTestId("textInput"), itemName);
  await act(async () => fireEvent(getByTestId("submit"), "onPress"));
  await act(async () => flushMicrotasksQueue());

  expect(minProps.navigation.navigate).not.toHaveBeenCalledWith();
});

test("should navigate to 'Search' screen", async () => {
  const expectedRoute = "Search";
  const expectedValue = { barcode: "123456789", type: "EAN" };

  const { getByTestId } = render(<ItemDetails {...minProps} />);

  await act(async () => fireEvent(getByTestId("search"), "onFocus"));

  expect(minProps.navigation.navigate).toHaveBeenCalledWith(
    expectedRoute,
    expectedValue
  );
});
