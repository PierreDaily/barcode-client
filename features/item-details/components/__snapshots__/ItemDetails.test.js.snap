// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Object {
      "display": "flex",
      "flex": 1,
      "justifyContent": "space-between",
      "marginLeft": 40,
      "marginRight": 40,
    }
  }
>
  <Logo
    style={
      Object {
        "alignSelf": "center",
        "marginTop": 61,
        "width": 500,
      }
    }
  />
  <View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(231, 231, 231)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          Object {
            "backgroundColor": "transparent",
            "marginBottom": 60,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.26)",
            "bottom": 0,
            "height": 2,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": Array [
              Object {
                "scaleY": 0.5,
              },
            ],
          }
        }
      />
      <View
        style={
          Object {
            "minHeight": 64,
            "paddingBottom": 0,
            "paddingTop": 0,
          }
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <Text
            numberOfLines={1}
            onLayout={[Function]}
            style={
              Object {
                "color": "#6200ee",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 12,
                "position": "absolute",
                "textAlign": "left",
                "top": 34,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            Product name
          </Text>
          <Text
            numberOfLines={1}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 12,
                "position": "absolute",
                "textAlign": "left",
                "top": 34,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            Product name
          </Text>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder=""
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#6200ee"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "textAlign": "left",
                "zIndex": 1,
              },
              Object {
                "paddingHorizontal": 12,
              },
              Object {
                "height": 64,
              },
              Object {
                "paddingBottom": 4,
                "paddingTop": 24,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlignVertical": "center",
              },
            ]
          }
          testID="textInput"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "rgb(231, 231, 231)",
            "borderTopLeftRadius": 4,
            "borderTopRightRadius": 4,
          },
          Object {
            "backgroundColor": "transparent",
            "marginBottom": 20,
          },
        ]
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "rgba(0, 0, 0, 0.26)",
            "bottom": 0,
            "height": 2,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "transform": Array [
              Object {
                "scaleY": 0.5,
              },
            ],
          }
        }
      />
      <View
        style={
          Object {
            "minHeight": 64,
            "paddingBottom": 0,
            "paddingTop": 0,
          }
        }
      >
        <View
          pointerEvents="none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "opacity": 1,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "transform": Array [
                Object {
                  "translateX": 0,
                },
              ],
            }
          }
        >
          <Text
            numberOfLines={1}
            onLayout={[Function]}
            style={
              Object {
                "color": "#6200ee",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 12,
                "position": "absolute",
                "textAlign": "left",
                "top": 34,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            Brand Name
          </Text>
          <Text
            numberOfLines={1}
            style={
              Object {
                "color": "rgba(0, 0, 0, 0.54)",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "left": 0,
                "opacity": 0,
                "paddingHorizontal": 12,
                "position": "absolute",
                "textAlign": "left",
                "top": 34,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateY": 0,
                  },
                  Object {
                    "scale": 1,
                  },
                ],
                "writingDirection": "ltr",
              }
            }
          >
            Brand Name
          </Text>
        </View>
        <TextInput
          allowFontScaling={true}
          editable={true}
          multiline={false}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder=""
          placeholderTextColor="rgba(0, 0, 0, 0.54)"
          rejectResponderTermination={true}
          selectionColor="#6200ee"
          style={
            Array [
              Object {
                "flexGrow": 1,
                "margin": 0,
                "textAlign": "left",
                "zIndex": 1,
              },
              Object {
                "paddingHorizontal": 12,
              },
              Object {
                "height": 64,
              },
              Object {
                "paddingBottom": 4,
                "paddingTop": 24,
              },
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": undefined,
                "textAlignVertical": "center",
              },
            ]
          }
          testID="search"
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#6200ee",
        "borderColor": "transparent",
        "borderRadius": 10,
        "borderStyle": "solid",
        "borderWidth": 0,
        "elevation": 2,
        "height": 40,
        "marginBottom": 46,
        "minWidth": 64,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 0.75,
          "width": 0,
        },
        "shadowOpacity": 0.24,
        "shadowRadius": 1.5,
      }
    }
    testID="submit"
  >
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 10,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "center",
            },
            Object {
              "borderRadius": 10,
              "height": 40,
            },
          ]
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "letterSpacing": 1,
                  "marginHorizontal": 16,
                  "marginVertical": 9,
                  "textAlign": "center",
                },
                undefined,
                Object {
                  "textTransform": "uppercase",
                },
                Object {
                  "color": "#ffffff",
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "fontFamily": "System",
                  "fontWeight": "500",
                },
                Object {
                  "color": "#F8F8F8",
                },
              ],
            ]
          }
        >
          Next
        </Text>
      </View>
    </View>
  </View>
</View>
`;
